### PR TITLE
Setup codingclub session

### PR DESCRIPTION
**Pull request by niconoe**
This is a work in progress, some tests are still lacking.

### CODE REVIEW

**Comment by ThierryO** [comment]
Please make sure that `R CMD check` passes on your machine. The CI fails on one of the vignettes.

**Comment by niconoe** [comment]
@ThierryO : indeed, `devtools::check()` is successful on my machine (`0 errors ✓ | 0 warnings ✓ | 1 note x`)

**Comment by ThierryO** [comment]
Pleas compare the packages on your machine with the ones installed by Travis

**Comment by damianooldoni** [comment]
Can somebody review this? @hansvancalster and @niconoe  are on holidays and I would like to have this function working before sending email with reminder and instructions for next coding club, which is planned on Aug, 25. I am planning to send the such email on next Thursday or Friday. Maybe @florisvdh  or @ThierryO? But anybody else is welcome, just run:

```r
devtools::install_github("inbo/inborutils@setup_codingclub_session")
library(inborutils)
?setup_codingclub_session() #get info about new functions and test it using examples first
```

Thanks everybody for any kind of suggestions.

File: R/setup_codingclub_session.R
Comment (florisvdh) [inline R/setup_codingclub_session.R]:
```suggestion
#'   not already present.
```

File: R/setup_codingclub_session.R
Comment (florisvdh) [inline R/setup_codingclub_session.R]:
```suggestion
#'   directories for a specific coding club session, the date of which is passed as a
```

File: R/setup_codingclub_session.R
Comment (florisvdh) [inline R/setup_codingclub_session.R]:
```suggestion
#'   parameter. Directories are created if needed. Files are downloaded if
```

**Review by florisvdh — APPROVED** [review]
It works @damianooldoni @niconoe !

I'll approve so that you can proceed - here are some remarks:

- is it expected that data go into folder `data/YYYYMMDD` and source files in `src/YYYYMMDD`? If so, that's indeed what it does :smile:
- if nothing is available for a certain date, the messages are a bit misleading. Yes, there are 2 warnings at the end, but above that it first says 'download completed'. Ideally at least the 'completed' lines are removed in that case:

```
* Download source file(s)...
* Download source file(s) completed
* Download data file(s)...
* Download data file(s) completed
Warning messages:
1: In value[[3L]](cond) :
  No content for this session (20200825) found. Is the date correct?
2: In value[[3L]](cond) :
  No content for this session (20200825) found. Is the date correct?
```

Regarding my suggested edits, don't forget to update the Rd file accordingly.

Below is my test session and resulting folder structure.

<details><summary>R code & output</summary>

```r
> remotes::install_github("inbo/inborutils@setup_codingclub_session")
Using github PAT from envvar GITHUB_PAT
Skipping install of 'inborutils' from a github remote, the SHA1 (883b81f3) has not changed since last install.
  Use `force = TRUE` to force installation
> library(inborutils)
> ?setup_codingclub_session()
starting httpd help server ... done
> getwd()
[1] "/home/floris"
> setwd("ContinuLegen/gooiweg")
> getwd()
[1] "/media/floris/DATA/ContinuLegen/gooiweg"
> setup_codingclub_session()
R scripts in https://github.com/inbo/coding-club/tree/master/data/20200820 will be downloaded in folder: ./src/20200820
Data in https://github.com/inbo/coding-club/tree/master/src/20200820 will be downloaded in folder: ./data/20200820
Do you want to continue? (Y/N) Y
* Download source file(s)...
* Download source file(s) completed
* Download data file(s)...
* Download data file(s) completed
Warning messages:
1: In value[[3L]](cond) :
  No content for this session (20200820) found. Is the date correct?
2: In value[[3L]](cond) :
  No content for this session (20200820) found. Is the date correct?
> setup_codingclub_session("20191126")
R scripts in https://github.com/inbo/coding-club/tree/master/data/20191126 will be downloaded in folder: ./src/20191126
Data in https://github.com/inbo/coding-club/tree/master/src/20191126 will be downloaded in folder: ./data/20191126
Do you want to continue? (Y/N) 
[1] "* Download aborted"
> setup_codingclub_session("20191126")
R scripts in https://github.com/inbo/coding-club/tree/master/data/20191126 will be downloaded in folder: ./src/20191126
Data in https://github.com/inbo/coding-club/tree/master/src/20191126 will be downloaded in folder: ./data/20191126
Do you want to continue? (Y/N) y
* Download source file(s)...
** Downloading https://github.com/inbo/coding-club/blob/master/src/20191126/20191126_challenges.R
** Downloading https://github.com/inbo/coding-club/blob/master/src/20191126/20191126_challenges_solution.R
* Download source file(s) completed
* Download data file(s)...
** Downloading https://github.com/inbo/coding-club/blob/master/data/20191126/20191126_BXL_iNaturalist_top20.csv
* Download data file(s) completed
> setup_codingclub_session("20200825", "cc_2020_08")
R scripts in https://github.com/inbo/coding-club/tree/master/data/20200825 will be downloaded in folder: cc_2020_08/src/20200825
Data in https://github.com/inbo/coding-club/tree/master/src/20200825 will be downloaded in folder: cc_2020_08/data/20200825
Do you want to continue? (Y/N) y
* Download source file(s)...
* Download source file(s) completed
* Download data file(s)...
* Download data file(s) completed
Warning messages:
1: In value[[3L]](cond) :
  No content for this session (20200825) found. Is the date correct?
2: In value[[3L]](cond) :
  No content for this session (20200825) found. Is the date correct?
> setup_codingclub_session("20200326", "cc_2020_03")
R scripts in https://github.com/inbo/coding-club/tree/master/data/20200326 will be downloaded in folder: cc_2020_03/src/20200326
Data in https://github.com/inbo/coding-club/tree/master/src/20200326 will be downloaded in folder: cc_2020_03/data/20200326
Do you want to continue? (Y/N) y
* Download source file(s)...
** Downloading https://github.com/inbo/coding-club/blob/master/src/20200326/20200326_challenges.R
** Downloading https://github.com/inbo/coding-club/blob/master/src/20200326/20200326_challenges_solutions.R
* Download source file(s) completed
* Download data file(s)...
** Downloading https://github.com/inbo/coding-club/blob/master/data/20200326/20200326_Belgian_municipalities.geojson
** Downloading https://github.com/inbo/coding-club/blob/master/data/20200326/20200326_butterflies.txt
** Downloading https://github.com/inbo/coding-club/blob/master/data/20200326/20200326_protected_areas.gpkg
** Downloading https://github.com/inbo/coding-club/blob/master/data/20200326/20200326_utm10_bel.gpkg
* Download data file(s) completed
```

</details>

<details><summary>folder structure</summary>

```bash
$ tree
.
├── cc_2020_03
│   ├── data
│   │   └── 20200326
│   │       ├── 20200326_Belgian_municipalities.geojson
│   │       ├── 20200326_butterflies.txt
│   │       ├── 20200326_protected_areas.gpkg
│   │       └── 20200326_utm10_bel.gpkg
│   └── src
│       └── 20200326
│           ├── 20200326_challenges.R
│           └── 20200326_challenges_solutions.R
├── data
│   └── 20191126
│       └── 20191126_BXL_iNaturalist_top20.csv
└── src
    └── 20191126
        ├── 20191126_challenges.R
        └── 20191126_challenges_solution.R

9 directories, 9 files
```

</details>

**Comment by damianooldoni** [comment]
Thanks @florisvdh for your inputs. 

1. documentation improved based on your  suggestions :+1:  (d2803ac)
2. yes, data go into folder `data/YYYYMMDD` and source files in `src/YYYYMMDD`
3. Error/warning/messages are definitely misleading, you are right! It can be better, for sure. I will work on it

For the coding club of next Tuesday we need just a R script (no data to download, scoop! :scream: ) so I think it is better to not hurry up. In this way @hansvancalster  and @niconoe will have time to further test so that we have tested on different machines: Windows 7 (me), Windows 10 (@hansvancalster, I think), a Mac OS X (@niconoe) and Linux (@florisvdh).

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: Solved: https://github.com/inbo/inborutils/commit/d2803ac2686eb543ba5333acafe10f9b9207cca9

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: https://github.com/inbo/inborutils/commit/d2803ac2686eb543ba5333acafe10f9b9207cca9

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: https://github.com/inbo/inborutils/commit/d2803ac2686eb543ba5333acafe10f9b9207cca9

**Comment by niconoe** [comment]
FYI, I installed a fresh R environment on Archlinux (all very recent packages: R 4.0.2, RStudio 1.3.1056, ...) and it seems to work well (manually using the helper, but also running `devtools::check()` and `devtools::test()`.

**Comment by niconoe** [comment]
Seems also fine on Mac OS (R 3.6.2, RStudio 1.2.5033)

**Comment by niconoe** [comment]
... And on Windows 10 with R 3.6.3 and RStudio 1.2.

I was just wondering if it wouldn't be nice if there was a default answer (yes) for the `Do you want to continue? (Y/N)` question. In practice: 

- the question string would be changed to `Do you want to continue? (Y/n)` (uppercase Y)
- an empty answer would be considered as a yes

Do like this change? If so, I can implement that.

**Comment by florisvdh** [comment]
I also favour having a default. Currently, it appears to be 'no' (from my earlier test session), but that's not indicated in the question. Would also propose 'Y' as default.

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R]:
```suggestion
    message("* Download aborted")
```

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R]: Consider an early return. It makes code more readable as it eliminates some of the curly branches.

```
if (continue != "y") {
 message("aborted")
 return(invisible(NULL))
}
do the stuff when continue == "y"
```

**Comment by damianooldoni** [comment]
No bad at all! Up to now Y and y are both valid options and anything else is treated like n.
@niconoe: I mocked the use of Y/y in tests by intensively using mockery package in addition of mocking the Sys.date function as you did already (double stabbing).

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R]: No message is case of a failure? Maybe move the message into `download_content_in_subdir()`.

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R]:
```suggestion
                                       github_subdirectory = c("src", "data") {
```

Then you can use `github_subdirectory <- match.arg(github_subdirectory)` instead of the assert_that()

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R]: I tend to use `dir.create(showWarnings = FALSE)`. Then the dir is created if it doesn't exists. Otherwise nothing happens. This would remove the if-else structure. Making it easier to create a unit test.

**Comment by florisvdh** [comment]
@damianooldoni I would keep the validity of both entered values Y and y as meaning 'yes'. I think @niconoe 's suggestion is only about the presentation of the question on screen (one as capital to denote the default - it is often done by interactive shell programs). The default is taken when nothing is entered (i.e. just pressing enter).

Actually I'm not sure what's the behaviour in other programs when entering some other random letter. I also rather think it should then be 'no' --> confirmed :wink: :

```bash
$ apt install gimp
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  gimp-data libamd2 libbabl-0.1-0 libcamd2 libccolamd2 libcholmod3 libgegl-0.4-0 libgegl-common libgexiv2-2 libgimp2.0 libheif1 libmetis5 libmypaint-1.5-1 libmypaint-common libraw19 libumfpack5
Suggested packages:
  gimp-help-en | gimp-help gimp-data-extras
The following NEW packages will be installed:
  gimp gimp-data libamd2 libbabl-0.1-0 libcamd2 libccolamd2 libcholmod3 libgegl-0.4-0 libgegl-common libgexiv2-2 libgimp2.0 libheif1 libmetis5 libmypaint-1.5-1 libmypaint-common libraw19
  libumfpack5
0 upgraded, 17 newly installed, 0 to remove and 8 not upgraded.
Need to get 15,3 MB of archives.
After this operation, 90,3 MB of additional disk space will be used.
Do you want to continue? [Y/n] t
Abort.
```

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R]: Vectorise the check what files already exist. Then loop only over the nonexisting files. This removes an if-else within a for loop -> huge simplification of the code.

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R:143]: Good candidate for an early return.

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R]:
```suggestion
  src_target_dir <- file.path(root_dir, src_rel_path, session_date)
```
This will use OS specific paths. Maybe add
```
src_target_dir <- normalizePath(src_target_dir)
```

File: R/setup_codingclub_session.R
Comment (ThierryO) [inline R/setup_codingclub_session.R]:
```suggestion
  data_target_dir <- file.path(root_dir, data_rel_path, session_date)
```

**Review by ThierryO — COMMENTED** [review]
Here are a few suggestion that would improve the code IMHO.

**Comment by damianooldoni** [comment]
Thanks @ThierryO for your valuable suggestions! I gave a fast look and as far as I see they will surely improve the code.

@niconoe : while working on Thierry's review, I can manage to take care about the default argument you proposed in https://github.com/inbo/inborutils/pull/85#issuecomment-680717737.

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: Totally agree. Solved.

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: Done: 435c9248f5f8101b5d6d82b74d53bb32aea00e30

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: Good idea. Done d9c4e35754fc0a419701169882ca18bf22711552

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: Inderdaad, done c59e6254720ed0d63e72f7a97c1112b01683f78c

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R:143]: Done 47b3613179b919e0f8988af891940d46ebc3ed6b

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: Nice idea: done, 24731b07857193aaf6e71d083956d20054acdeb0

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: Done in same commit 24731b07857193aaf6e71d083956d20054acdeb0

File: R/setup_codingclub_session.R
Comment (damianooldoni) [reply R/setup_codingclub_session.R]: In case of failuer error message is returned. But you are right, the message should be moved into `download_content_in_subdir()`. In this way I have not to write twice almost same message, but I can use value of argument `github_subdirectory` in message.

Done 4a532017f4d0f5d34372c775c78e811aa2e96394

**Comment by damianooldoni** [comment]
@hansvancalster, as coding club core team member, I think you have the right to have the last look on this PR before merging. Thanks a lot. But if anybody else wants to add something, feel free to shoot it. Thanks in advance for the interesting reviews all of you gave to this function.

**Comment by hansvancalster** [comment]
@damianooldoni sorry to not have looked at it. I just haven't find the time yet. But I'm very confident this will work just fine as @niconoe and you developed it and @ThierryO and @florisvdh have thoroughly reviewed it.
So the honor is all yours. Go ahead and merge this branch :+1:  Thanks very much to all for having implemented this :rocket:

**Review by hansvancalster — APPROVED** [review]
see other comment

**Pull request merged**
